pkg/staking/stakingcontract/contract.go

```diff
@@ -0,0 +1,202 @@
+	GetStake(ctx context.Context, overlay swarm.Address) (big.Int, error)
+}
+
+type contract struct {
+	owner                  common.Address
+	stakingContractAddress common.Address
+	bzzTokenAddress        common.Address
+	transactionService     transaction.Service
+}
+
+func New(
+	owner common.Address,
+	stakingContractAddress common.Address,
+	bzzTokenAddress common.Address,
+	transactionService transaction.Service,
+) Contract {
+	return &contract{
+		owner:                  owner,
+		stakingContractAddress: stakingContractAddress,
+		bzzTokenAddress:        bzzTokenAddress,
+		transactionService:     transactionService,
+	}
+}
+
+func (s *contract) sendTransaction(ctx context.Context, callData []byte, desc string) (*types.Receipt, error) {
+	request := &transaction.TxRequest{
+		To:          &s.stakingContractAddress,
+		Data:        callData,
+		GasPrice:    sctx.GetGasPrice(ctx),
+		GasLimit:    1600000,
+		Value:       big.NewInt(0),
+		Description: desc,
+	}
+
+	txHash, err := s.transactionService.Send(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := s.transactionService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if receipt.Status == 0 {
+		return nil, transaction.ErrTransactionReverted
+	}
+
+	return receipt, nil
+}
+
+func (s *contract) sendDepositStakeTransaction(ctx context.Context, owner common.Address, stakedAmount big.Int, nonce common.Hash) (*types.Receipt, error) {
+
+	callData, err := stakingABI.Pack("depositStake", owner, stakedAmount, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := s.sendTransaction(ctx, callData, depositStakeDescription)
+	if err != nil {
+		return nil, fmt.Errorf("deposit stake: stakedAmount %d: %w", &stakedAmount, err)
+	}
+
+	return receipt, nil
+}
+
+func (s *contract) sendGetStakeTransaction(ctx context.Context, overlay []byte) (*big.Int, error) {
+
+	callData, err := stakingABI.Pack("stakeOfOverlay", overlay)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := s.transactionService.Call(ctx, &transaction.TxRequest{
+		To:   &s.stakingContractAddress,
+		Data: callData,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("get stake: overlayAddress %d: %w", overlay, err)
+	}
+
+	results, err := erc20ABI.Unpack("stakeOfOverlay", result)
+	if err != nil {
+		return nil, err
+	}
+	return abi.ConvertType(results[0], new(big.Int)).(*big.Int), nil
+}
+
+func (s *contract) DepositStake(ctx context.Context, stakedAmount big.Int, overlay swarm.Address) error {
+	prevStakedAmount, err := s.sendGetStakeTransaction(ctx, overlay.Bytes())
+	if err != nil {
+		return err
+	}
+
+	if prevStakedAmount.Cmp(big.NewInt(0)) == -1 {
+		if stakedAmount.Cmp(MinimumStakeAmount) == -1 {
+			return ErrInvalidStakeAmount
+		}
+	}
+
+	balance, err := s.getBalance(ctx)
+	if err != nil {
+		return err
+	}
+
+	if balance.Cmp(&stakedAmount) <= 0 {
+		return ErrInsufficientFunds
+	}
```

Comment:
`balance.Cmp(&stakedAmount) <= 0` 

what if user wants to stake entire balance? do we want to forbid this?

Reply:
i was thinking maybe some bzz would be used for transaction (maybe gas fee?) which is why i implemented it this way.

Reply:
1) gas fees should come from native token of network
2) even with current logic we still have problem if gas fees are applied to bzz balance (case: balance is 20bzz, stake amount is 19bzz, fees are 2bzz).

based on both points we should allow stake amount to be up to and including entire balance